tratto/src/main/java/star/tratto/util/javaparser/JavaParserUtils.java

```diff
@@ -471,52 +656,189 @@ public static String getMethodSignature(MethodDeclaration methodDeclaration) {
-        Matcher matcher = METHOD_SIGNATURE.matcher(methodDeclaration);
+    public static String getMethodSignature(
+            MethodUsage methodUsage
+    ) {
+        ResolvedMethodDeclaration methodDeclaration = methodUsage.getDeclaration();
+        // Consider JavaParserMethodDeclaration.
+        if (methodDeclaration.toString().startsWith("JavaParserMethodDeclaration")) {
+            JavaParserMethodDeclaration jpMethodDeclaration = (JavaParserMethodDeclaration) methodDeclaration;
+            MethodDeclaration jpMethod = jpMethodDeclaration.getWrappedNode();
+            return getMethodSignature(jpMethod);
+        }
+        // Consider ReflectionMethodDeclaration or JavassistMethodDeclaration
+        Matcher matcher = METHOD_SIGNATURE.matcher(methodDeclaration.toString());
         if (!matcher.find()) {
             throw new IllegalStateException("Could not parse method signature: " + methodDeclaration);
         }
-        String methodModifiers = methodDeclaration.startsWith("ReflectionMethodDeclaration") ? matcher.group(1) : matcher.group(2);
-        // Take into account the case in which the method declaration refers to a method without an access specifier
+        String methodModifiers = methodDeclaration.toString().startsWith("ReflectionMethodDeclaration") ? matcher.group(1) : matcher.group(2);
+        // Take into account the case in which the method declaration refers to a method without an access specifier.
         if (methodModifiers == null) {
-            assert methodDeclaration.startsWith("ReflectionMethodDeclaration");
+            assert methodDeclaration.toString().startsWith("ReflectionMethodDeclaration");
             methodModifiers = "";
         }
-        List<String> methodTypeParameters = new ArrayList<>();
-        for (int i=0; i < methodUsage.getDeclaration().getTypeParameters().size(); i++) {
-            methodTypeParameters.add(methodUsage.getDeclaration().getTypeParameters().get(i).getName());
+        List<String> typeParameterList = getMethodUsageTypeParameters(methodUsage);
+        List<String> parameterList = getMethodUsageParameters(methodUsage);
+        List<String> exceptionList = getMethodUsageExceptions(methodUsage);
+        return (methodModifiers + " " + (typeParameterList.isEmpty() ? "" : "<" + String.join(", ", typeParameterList) + ">") +
+                " " + getTypeWithoutPackages(methodDeclaration.getReturnType()) +
+                " " + methodDeclaration.getName() +
+                "(" + String.join(", ", parameterList) + ")" +
+                (exceptionList.isEmpty() ? "" : " throws " + String.join(", ", exceptionList)))
+                .replaceAll(" +", " ").trim();
+    }
+
+    /**
+     * Gets the package of a given compilation unit.
+     *
+     * @param cu a compilation unit
+     * @return the package of the compilation unit
+     * @throws PackageDeclarationNotFoundException if the package declaration
```

Comment:
Here, "package declaration" refers to a JavaParser concept rather than a Java concept, so I would indicate that explicitly.

What does this method do if the package is the unnamed package?  I presume it counts as being found, and would always be found, but I can't tell from the method documentation.

Reply:
Currently, due to the nature of JavaParser, I believe that the method will throw `PackageDeclarationNotFoundException` if the package is unnamed. What would be a good alternative? Is there another way to retrieve the unnamed package? My only concern is that we may violate a few scoping restrictions when generating the tokens dataset.